### PR TITLE
UI: Ensure there will be no race condition when getting the UI_METHOD ex_data

### PR DESCRIPTION
**Pull request by levitte**
##### Checklist

- [x] CLA is signed

### CODE REVIEW

**Comment by levitte** [comment]
@kaduk, I think this interests you.

**Comment by kaduk** [comment]
It does interest me, thanks :)
Though, it looks like we end up using the index of 0 if the RUN_ONCE fails, which is maybe not desired.
(Should RUN_ONCE return -1 instead of 0 for failure?)

**Comment by levitte** [comment]
> Though, it looks like we end up using the index of 0 if the RUN_ONCE fails, which is maybe not desired.
(Should RUN_ONCE return -1 instead of 0 for failure?)

I'm taking advantage of a small quirk in `CRYPTO_get_ex_new_index()`...  from the manual:

```
       CRYPTO_get_ex_new_index() returns a new index or -1 on failure; the
       value 0 is reserved for the legacy "app_data" API's.
```

So, in essence, it will never return 0.

**Comment by kaduk** [comment]
CRYPTO_get_ex_new_index() will never return 0, yes, but CRYPTO_THREAD_run_once() might [e.g., if the pthread library is busted, so it's unlikely, but allowed by the API contract].  (I'm looking at the RUN_ONCE() macro.)

**Comment by levitte** [comment]
D'oh.  Of course you're right.  I just pushed a fix

**Comment by richsalz** [comment]
LGTM.  +1 after @kaduk likes it :)

File: crypto/ui/ui_util.c
Comment (kaduk) [inline crypto/ui/ui_util.c]: Does this one need to check the return value, too?

Comment (levitte) [inline crypto/ui/ui_util.c]: I don't think so.  The call in `UI_UTIL_wrap_read_pem_callback()` will always happen first, is there a reason to believe that a later run of the exact same values will trigger a fault?

Comment (levitte) [inline crypto/ui/ui_util.c]: @mattcaswell might be the guy to answer this...

Comment (richsalz) [inline crypto/ui/ui_util.c]: on second thought... that embedded RUN_ONCE call is a bit tricky ...

Comment (levitte) [inline crypto/ui/ui_util.c]: > that embedded RUN_ONCE call is a bit tricky ...

How so?

Comment (kaduk) [inline crypto/ui/ui_util.c]: > The call in ```UI_UTIL_wrap_read_pem_callback()``` will always happen first

I sort-of suspected that was the case, but I'm still a little foggy after being out sick yesterday, so wanted to double-check.

I guess we should let Rich resolve his "a bit tricky" remark before you merge it, though.

Comment (richsalz) [inline crypto/ui/ui_util.c]: it's a macro calling function exactly once with a side-effect that returns a magic file-static variable.  why not just put that RUN_ONCE call as a separate statement?  or perhaps have a global variable and init in the exdata run-once block?

Comment (levitte) [inline crypto/ui/ui_util.c]: So essentially, you want me to add *yet another* file-static variable (there's not reason it should be more global than that) and add one more call level?

Comment (richsalz) [inline crypto/ui/ui_util.c]: yes, but if you object i won't block this.  the other places with index do it that way as i recall.

Comment (levitte) [inline crypto/ui/ui_util.c]: > the other places with index do it that way as i recall.

Yes. They often create objects that can't be returned as int or initialize stuff that don't have a sensible value to return...

Comment (richsalz) [inline crypto/ui/ui_util.c]: they return a status code.  Returning a value isn't common.  a file-static with the index is cleaner.   IMO.  If you disagree, I won't fight it.

Comment (levitte) [inline crypto/ui/ui_util.c]: Okie.  I consider this questions resolved then.

**Comment by levitte** [comment]
So after the discussion on the use of RUN_ONCE, where do we stand?

**Comment by mattcaswell** [comment]
> So, in essence, it will never return 0.

But the RUN_ONCE macro *can* return 0 on failure. Look at its definition. It is documented in the internal header as follows:
````
/*
 * RUN_ONCE - use CRYPTO_THREAD_run_once, and check if the init succeeded
 * @once: pointer to static object of type CRYPTO_ONCE
 * @init: function name that was previously given to DEFINE_RUN_ONCE,
 *        DEFINE_RUN_ONCE_STATIC or DECLARE_RUN_ONCE.
 *
 * The return value is 1 on success or 0 in case of error.
 */
````

I don't like the use of `RUN_ONCE` here at all. You should keep it to returning just 1 or 0 to be consistent with all the other uses. Otherwise this turns into a big gotcha for future code maintainers. I agree with @richsalz on this.

**Comment by levitte** [comment]
That leads to something I see as a problem, and it's giving the rest of the code access to a very private variable that shouldn't be tampered with.  When protecting such variables, I usually code like this:

```
type fn()
{
    static type privatevar = -1;  /* that value is just an example */

    if (privatevar == -1) {
        /* code that initialises privatevar */
    }

    return privatevar;
}
```

That way, there's only one way `privatevar` can be accessed, but most of all, tampered with.

What you are forcing me to do, as far as I can see, is either to make a variable that's openly accessible to the rest of the file, or create a function like the one above that might or might not be called before RUN_ONCE has been performed.  I see that as a coding safety issue.

**Comment by levitte** [comment]
Granted, the "magic variable" that RUN_ONCE uses is openly available as well...  but at the very least, it's hidden away.

**Comment by levitte** [comment]
I have added another commit, which should please @richsalz and @mattcaswell a bit more.

**Review by mattcaswell — APPROVED** [review]

**Comment by levitte** [comment]
Merged

**Pull request closed without merging**
